packages/flutter/lib/src/scheduler/binding.dart

```diff
@@ -436,8 +436,9 @@ mixin SchedulerBinding on BindingBase {
   // Scheduled by _ensureEventLoopCallback.
   void _runTasks() {
     _hasRequestedAnEventLoopCallback = false;
-    if (handleEventLoopCallback())
-      _ensureEventLoopCallback(); // runs next task when there's time
+    if (handleEventLoopCallback() || _taskQueue.isNotEmpty) {
```

Comment:
It looks like this can be simplified since there isn't really a case where `handleEventLoopCallback` returns true and `_taskQueue.isNotEmpty` is false, which means only `_taskQueue.isNotEmpty` decides here whether we queue up another event loop callback.

Reply:
Actually, this doesn't look correct: If `handleEventLoopCallback` returns false because we're `locked` (see line 457) we are not allowed to call `_ensureEventLoopCallback` because of the `assert(!locked)` in that method.

Reply:
Yeah, I also think it  doesn't look correct with the previous change. After I thought it over and over again, I think the problem is inside the `handleEventLoopCallback`. 
Because 
![image](https://user-images.githubusercontent.com/40540394/126061374-bb057783-32e2-4bb0-95ab-65eb73db3332.png)


According to the source code, The `handleEventLoopCallback` returns false in three cases：
*  _taskQueue.isEmpty || locked 
*  `schedulingStrategy()` return true, and `_taskQueue` becomes empty after performing a task 
*  **`schedulingStrategy()` return false** , although _taskQueue is not empty

I think the third case is problematic, because going to the third logic says the `_taskQueue` is not empty and  `locked` is false . This indicates that a task still needs to be executed, except that this time the `schedulingStrategy` is not satisfied. It should return true to continue the  `_ensureEventLoopCallback`  like this.

![image](https://user-images.githubusercontent.com/40540394/126061541-281437b1-56ca-444a-b2af-cfa766109673.png)

Reply:
This seems incorrect to me, since it can cause tasks to fire even if the `locked` is true.

Reply:
There was something wrong with my previous revision, you can check out my latest comment .
https://github.com/flutter/flutter/pull/82781/files#r671808835

I will revert the previous changes. And I think the problem is just that `handleEventLoopCallback `.

If locked is true, this method will return fasle at the first step and the event will not be executed
![image](https://user-images.githubusercontent.com/40540394/129759747-2dd83280-a62f-40ea-a39a-7b12416b2269.png)